File: example/kaggle-ndsb2/preprocessing.R
unzip_data <- function(root_path){
  stopifnot(file.exists("train.zip"))
  stopifnot(file.exists("validate.zip"))
  
  if (!dir.exists(root_path))
    dir.create(root_path)
  
  unzip("sample_submission_validate.csv.zip", exdir = root_path)
  unzip("train.csv.zip", exdir = root_path)
  
  if (!dir.exists(paste0(root_path, "/train")))
    unzip("train.zip", exdir = root_path)
  if (!dir.exists(paste0(root_path, "/validate"))) {
    unzip("validate.zip", exdir = root_path)
  }
}

get_frames <- function(root_path){
  list.files(root_path, full.names = TRUE) %>>% file.info %>>% (rownames(.)[.$isdir]) %>>% 
    lapply(function(subfolder){
      list.dirs(subfolder, TRUE, TRUE) %>>% `[`(str_detect(., "sax_\\d+$")) %>>%
        lapply(function(subfolder2){
          list.files(subfolder2, "-\\d{4}.dcm$", recursive = TRUE, full.names = TRUE)
        })
    }) %>>% do.call(what = c)
}

write_label_csv <- function(fname, frames, label_map_file = NULL) {
  
  index <- sapply(frames, function(x) sapply(str_split(x, "/"), `[`, 3)) %>>% as.integer
  if (is.null(label_map_file)) {
    fwrite(data.table(index, 0, 0), fname, col.names = FALSE)
  } else {
    fwrite(fread(label_map_file)[index, ], fname, col.names = FALSE)
  }
  invisible(TRUE)
}

encode_csv <- function(label_csv, systole_csv, diastole_csv) {
  labelData <- fread(label_csv)
  systole_encode <- sapply(labelData$V2, `<`, 1:600) %>>% (matrix(as.integer(.), nrow(.))) %>>% t
  diastole_encode <- sapply(labelData$V3, `<`, 1:600) %>>% (matrix(as.integer(.), nrow(.))) %>>% t
  fwrite(data.table(systole_encode), systole_csv, col.names = FALSE)
  fwrite(data.table(diastole_encode), diastole_csv, col.names = FALSE)
  invisible(TRUE)
}

resizeImage <- function(image, targetSize) {
  minDim <- min(dim(image))
  stPixel <- (dim(image) - minDim) / 2 + 1
  tmp <- image[stPixel[1]:(minDim + stPixel[1]-1) , stPixel[2]:(minDim + stPixel[2]-1)]
  
  outGird <- mapply(function(ts, os) (1:ts - 0.5) * (os / ts) - 0.5, 
                    targetSize, dim(tmp), SIMPLIFY = FALSE)
  resultImg <- interp2(0:(nrow(tmp)-1), 0:(ncol(tmp)-1), tmp, outGird[[1]], outGird[[2]]) %>>%
    `*`(255) %>>% as.integer %>>% matrix(targetSize[1])
  return(resultImg)
}

write_data_csv <- function(fname, frames, preproc) {
  clusterExport(cl, "preproc", environment())
  data <- parLapply(cl, frames, function(path){
    lapply(path, function(imgFile){
      img <- readDICOMFile(imgFile)$img %>>% `[`(rev(1:nrow(.)), 1:ncol(.))
      if (diff(dim(img)) < 0) img <- t(img)
      as.vector(preproc(img / max(img)))
    }) %>>% do.call(what = c)
  })
  fwrite(data.table(do.call(rbind, data)), fname, col.names = FALSE)
  invisible(TRUE)
}

library(pipeR)
library(stringr)
library(parallel)
library(data.table)

# unzip train.zip and validate.zip
unzip_data("data")

# Load the list of all the training frames, and shuffle them
# Shuffle the training frames
set.seed(10)
train_frames <- get_frames("data/train") %>>% `[`(sample.int(length(.), length(.)))
validate_frames <- get_frames("data/validate") %>>% `[`(sample.int(length(.), length(.)))

# Write the corresponding label information of each frame into file.
write_label_csv("train-label.csv", train_frames, "data/train.csv")
write_label_csv("validate-label.csv", validate_frames)

# Write encoded label into the target csv
# We use CSV so that not all data need to sit into memory
# You can also use inmemory numpy array if your machine is large enough
encode_csv("train-label.csv", "train-systole.csv", "train-diastole.csv")

# open cluster for parallel processing
cl <- makeCluster(detectCores() - 1L)
tmpOutput <- clusterEvalQ(cl, {
  library(oro.dicom)
  library(data.table)
  library(pipeR)
  library(Rcpp)
  library(RcppArmadillo)
  Sys.setenv(PKG_CPP_FLAG = "-Icpp")
  sourceCpp("cpp/exportFuncs.cpp")
})
clusterExport(cl, "resizeImage")

# Dump the data of each frame into a CSV file, apply crop to 64 preprocessor
write_data_csv("train-64x64-data.csv", train_frames, function(img) resizeImage(img, c(64, 64)))
write_data_csv("validate-64x64-data.csv", train_frames, function(img) resizeImage(img, c(64, 64)))

# stop cluster
stopCluster(cl)

File: example/kaggle-ndsb2/Preprocessing.py
"""Preprocessing script.

This script walks over the directories and dump the frames into a csv file
"""
import os
import csv
import sys
import random
import scipy
import numpy as np
import dicom
from skimage import io, transform
from joblib import Parallel, delayed
import dill

def mkdir(fname):
   try:
       os.mkdir(fname)
   except:
       pass

def get_frames(root_path):
   """Get path to all the frame in view SAX and contain complete frames"""
   ret = []
   for root, _, files in os.walk(root_path):
       root=root.replace('\\','/')
       files=[s for s in files if ".dcm" in s]
       if len(files) == 0 or not files[0].endswith(".dcm") or root.find("sax") == -1:
           continue
       prefix = files[0].rsplit('-', 1)[0]
       fileset = set(files)
       expected = ["%s-%04d.dcm" % (prefix, i + 1) for i in range(30)]
       if all(x in fileset for x in expected):
           ret.append([root + "/" + x for x in expected])
   # sort for reproduciblity
   return sorted(ret, key = lambda x: x[0])


def get_label_map(fname):
   labelmap = {}
   fi = open(fname)
   fi.readline()
   for line in fi:
       arr = line.split(',')
       labelmap[int(arr[0])] = line
   return labelmap


def write_label_csv(fname, frames, label_map):
   fo = open(fname, "w")
   for lst in frames:
       index = int(lst[0].split("/")[3])
       if label_map is not None:
           fo.write(label_map[index])
       else:
           fo.write("%d,0,0\n" % index)
   fo.close()


def get_data(lst,preproc):
   data = []
   result = []
   for path in lst:
       f = dicom.read_file(path)
       img = preproc(f.pixel_array.astype(float) / np.max(f.pixel_array))
       dst_path = path.rsplit(".", 1)[0] + ".64x64.jpg"
       scipy.misc.imsave(dst_path, img)
       result.append(dst_path)
       data.append(img)
   data = np.array(data, dtype=np.uint8)
   data = data.reshape(data.size)
   data = np.array(data, dtype=np.str_)
   data = data.reshape(data.size)
   return [data,result]


def write_data_csv(fname, frames, preproc):
   """Write data to csv file"""
   fdata = open(fname, "w")
   dr = Parallel()(delayed(get_data)(lst,preproc) for lst in frames)
   data,result = zip(*dr)
   for entry in data:
      fdata.write(','.join(entry)+'\r\n')
   print("All finished, %d slices in total" % len(data))
   fdata.close()
   result = np.ravel(result)
   return result


def crop_resize(img, size):
   """crop center and resize"""
   if img.shape[0] < img.shape[1]:
       img = img.T
   # we crop image from center
   short_egde = min(img.shape[:2])
   yy = int((img.shape[0] - short_egde) / 2)
   xx = int((img.shape[1] - short_egde) / 2)
   crop_img = img[yy : yy + short_egde, xx : xx + short_egde]
   # resize to 64, 64
   resized_img = transform.resize(crop_img, (size, size))
   resized_img *= 255
   return resized_img.astype("uint8")


def local_split(train_index):
   random.seed(0)
   train_index = set(train_index)
   all_index = sorted(train_index)
   num_test = int(len(all_index) / 3)
   random.shuffle(all_index)
   train_set = set(all_index[num_test:])
   test_set = set(all_index[:num_test])
   return train_set, test_set


def split_csv(src_csv, split_to_train, train_csv, test_csv):
   ftrain = open(train_csv, "w")
   ftest = open(test_csv, "w")
   cnt = 0
   for l in open(src_csv):
       if split_to_train[cnt]:
           ftrain.write(l)
       else:
           ftest.write(l)
       cnt = cnt + 1
   ftrain.close()
   ftest.close()

# Load the list of all the training frames, and shuffle them
# Shuffle the training frames
random.seed(10)
train_frames = get_frames("./data/train")
random.shuffle(train_frames)
validate_frames = get_frames("./data/validate")

# Write the corresponding label information of each frame into file.
write_label_csv("./train-label.csv", train_frames, get_label_map("./data/train.csv"))
write_label_csv("./validate-label.csv", validate_frames, None)

# Dump the data of each frame into a CSV file, apply crop to 64 preprocessor
train_lst = write_data_csv("./train-64x64-data.csv", train_frames, lambda x: crop_resize(x, 64))
valid_lst = write_data_csv("./validate-64x64-data.csv", validate_frames, lambda x: crop_resize(x, 64))

# Generate local train/test split, which you could use to tune your model locally.
train_index = np.loadtxt("./train-label.csv", delimiter=",")[:,0].astype("int")
train_set, test_set = local_split(train_index)
split_to_train = [x in train_set for x in train_index]
split_csv("./train-label.csv", split_to_train, "./local_train-label.csv", "./local_test-label.csv")
split_csv("./train-64x64-data.csv", split_to_train, "./local_train-64x64-data.csv", "./local_test-64x64-data.csv")
